hazelcast/src/main/java/com/hazelcast/internal/config/AbstractConfigLocator.java

```diff
@@ -158,6 +158,14 @@ protected boolean loadConfigurationFromClasspath(String configFileName) {
             LOGGER.info(String.format("Loading '%s' from the classpath.", configFileName));
 
             configurationUrl = url;
+
+            // configurationFile must be set correctly because dynamic
+            // configuration persistence depends on this field. If this is
+            // absent, hazelcast instance may fail to find a file to persist.
+            File file = new File(url.getPath());
+            if (file.exists()) {
+                configurationFile = file;
+            }
```

Comment:
Please add a comment that this handles the `hazelcast.[xml|yaml|yml]` file on classpath, but only when the file is a regular file on the filesystem - e.g. hazelcast.xml in resources folder or directly lib directory and that it doesn't handle file inside a jar.

Reply:
https://github.com/hazelcast/hazelcast/pull/20399/commits/87461e12a333223d5a9a4175138136d6740fa4b0